src/DocumentDB.ChangeFeedProcessor/PartitionControllerHealthnessEvaluator.cs

```diff
@@ -0,0 +1,60 @@
+﻿//----------------------------------------------------------------
+// Copyright (c) Microsoft Corporation.  Licensed under the MIT license.
+//----------------------------------------------------------------
+
+namespace Microsoft.Azure.Documents.ChangeFeedProcessor
+{
+    using System;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using Microsoft.Azure.Documents.ChangeFeedProcessor.PartitionManagement;
+
+    internal class PartitionControllerHealthnessEvaluator : IPartitionController
+    {
+        private static readonly long UninitializedTicks = DateTime.MaxValue.Ticks;
+        private readonly IPartitionController inner;
+        private readonly long maxAllowedUnhealthyDuration;
+        private readonly IUnhealthyHandlingStrategy unhealthyStrategy;
+        private long firstUnhealthyTick = UninitializedTicks;
+
+        public PartitionControllerHealthnessEvaluator(IPartitionController inner, TimeSpan leaseExpirationTimeout, IUnhealthyHandlingStrategy unhealthyStrategy)
```

Comment:
This should take just the expiration duration/shouldn't know about leases, etc, the caller should be responsible to use lease expiration, etc.

Reply:
but it wraps the calls around PartitionController -> the whole point is to decorate IPartitionController.AquireLease -> if the controller is not able to aquire it, then it's useless and unhealthy